src/common/api/emote-menu.js

```diff
@@ -0,0 +1,98 @@
+/* eslint-disable import/prefer-default-export */
+// import {EmoteCategories, EmoteProviders} from '../../constants.js';
+import {EmoteCategories, EmoteProviders} from '../../constants.js';
+import Emote from '../../modules/emotes/emote.js';
+import emoteMenuViewStore from '../stores/emote-menu-view-store.js';
+
+const RESERVED_EMOTE_PROVIDER_IDS = Object.values(EmoteProviders);
+const RESERVED_EMOTE_CATEGORY_IDS = Object.values(EmoteCategories);
+
+export default {
+  upsertProvider({id, provider, displayName, icon, emotes: providerEmotes = [], channel = null}) {
+    if (id == null || id.length === 0) {
+      throw new Error('Category ID must be a non-empty string');
+    }
+
+    if (provider == null || provider.length === 0) {
+      throw new Error('Provider ID must be a non-empty string');
+    }
+
+    if (displayName == null || displayName.length === 0) {
+      throw new Error('Display name must be a non-empty string');
+    }
+
+    if (channel != null && channel.length === 0) {
+      throw new Error('Channel must be a non-empty string');
+    }
+
+    if (RESERVED_EMOTE_CATEGORY_IDS.includes(id)) {
```

Comment:
why do we permit providers to generate their own category identifiers? would it not be more wise to assign those and return them on creation? we could generate a uuid and return that. future changes would require the uuid be specified to update/delete.